pickhardtpayments/UncertaintyChannel.py

```diff
@@ -244,33 +255,42 @@ def get_piecewise_linearized_costs(self,number_of_pieces : int = DEFAULT_N,
 
         #using certainly available liquidity costs us nothing but fees
         if int((self.min_liquidity-self.in_flight)/quantization) > 0:
-            uncertintay_unit_cost = 0 #is zero as we have no uncertainty in this case!
+            uncertainty_unit_cost = 0 #is zero as we have no uncertainty in this case!
             pieces.append((int((self.min_liquidity-self.in_flight)/quantization),uncertintay_unit_cost + mu * self.linearized_integer_routing_unit_cost()))
             number_of_pieces-=1
 
         # FIXME: include the in_flight stuff
         if int(self.conditional_capacity/quantization) > 0 and number_of_pieces > 0:
             capacity = int(self.conditional_capacity/(number_of_pieces*quantization))
-            uncertintay_unit_cost = self.linearized_integer_uncertainty_unit_cost()
+            uncertainty_unit_cost = self.linearized_integer_uncertainty_unit_cost()
             for i in range(number_of_pieces):
-                a = (i+1)*uncertintay_unit_cost+1
+                a = (i+1)*uncertainty_unit_cost+1
                 b = self.linearized_integer_routing_unit_cost()+1
                 pieces.append((capacity, int(a*b/(a+mu*b)) ))
         return pieces
     """
 
-    def update_knowledge(self, amt: int, success_of_probe):
+    def update_knowledge(self, amt: int, probing_successful):
         """
         updates our knowledge about the channel if we tried to probe it for amount `amt`
 
         This API works only if we have an Oracle that allows to ask the actual liquidity of a channel
         In mainnet Lightning our oracle will not work on a per_channel level. This will change the data
         flow. Here for simplicity of the simulation we make use of the Oracle on a per channel level
         """
-        if success_of_probe:
-            self.min_liquidity = max(self.min_liquidity, self.in_flight+amt)
+        if probing_successful:
+            # self.min_liquidity = max(self.min_liquidity, self.in_flight+amt)
+            self.min_liquidity = max(self.min_liquidity, amt)
+            if self.return_channel:
+                self.return_channel.max_liquidity = min(self.return_channel.max_liquidity,
+                                                        self.return_channel.capacity - amt)
```

Comment:
or even `max(0,self.renturn_channel.max_liquidity - amt)`

Reply:
I was quite conservative on this one, I must admit. My thinking was along the lines:
"ok, what do I know now. I know the amount, that I can send through the channel in oracle network. meaning the back channel must be capacity minus the amount. But I do have a previous estimate of the maximum liquidity. Now, given that this is an estimate, should I reduce the amount from the estimate? is this then still a maximum? 

But I see your point. following this scheme, with the last payment that goes through, probably then a small one, I more and more increase the maximum again. basically I forget previous learning. which would be wrong.
So I agree, only using the previous max would incorporate previous learning.

Reply:
As a consequence, the call to update_knowledge in the payment.execute() method needs to be removed. 
Also, in case of failure of send_onion the liquidity estimates need to be moved back.